src/Datadog.NativeLoader/class_factory.cpp

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full
+// license information.
+#include "class_factory.h"
+
+#include "cor_profiler.h"
+#include "logging.h"
+
+const IID IID_IUnknown = {0x00000000, 0x0000, 0x0000, {0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46}};
+const IID IID_IClassFactory = {0x00000001, 0x0000, 0x0000, {0xC0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46}};
+
+ClassFactory::ClassFactory(datadog::nativeloader::DynamicDispatcher* dispatcher) : refCount(0), dispatcher(dispatcher)
+{
+    Debug("ClassFactory::.ctor");
+}
+
+ClassFactory::~ClassFactory()
+{
+}
+
+HRESULT STDMETHODCALLTYPE ClassFactory::QueryInterface(REFIID riid, void** ppvObject)
+{
+    Debug("ClassFactory::QueryInterface");
+    HRESULT res = dispatcher->LoadClassFactory(riid);
+
+    if ((riid == IID_IUnknown || riid == IID_IClassFactory) && SUCCEEDED(res))
+    {
+        *ppvObject = this;
+        this->AddRef();
+
+        return S_OK;
+    }
+
+    *ppvObject = nullptr;
+    return E_NOINTERFACE;
+}
+
+ULONG STDMETHODCALLTYPE ClassFactory::AddRef()
+{
+    Debug("ClassFactory::AddRef");
+    return std::atomic_fetch_add(&this->refCount, 1) + 1;
+}
+
+ULONG STDMETHODCALLTYPE ClassFactory::Release()
+{
+    Debug("ClassFactory::Release");
+    int count = std::atomic_fetch_sub(&this->refCount, 1) - 1;
+    if (count <= 0)
+    {
+        delete this;
+    }
+
+    return count;
+}
+
+// profiler entry point
+HRESULT STDMETHODCALLTYPE ClassFactory::CreateInstance(IUnknown* pUnkOuter, REFIID riid, void** ppvObject)
+{
+    Debug("ClassFactory::CreateInstance");
+    if (pUnkOuter != nullptr)
+    {
+        *ppvObject = nullptr;
+        return CLASS_E_NOAGGREGATION;
+    }
+
+    auto profiler = new datadog::nativeloader::CorProfiler(dispatcher);
```

Comment:
out of curiosity: shouldn't we keep this pointer and delete it in the dtor (or use a std::unique_ptr) ?

Reply:
I guess you're talking about the dispatcher... you're probably right, but not critical at this moment, due the life cycle of the dispatcher is until process finalization. Maybe I should delete the instance in the DllMain on detach instead using the unique_ptr? WDYT?

Reply:
yeah, that was mainly a question. We have same thing in the profiler and I was curious about you point of view.
As you said, it's not critical, we can come back to it later.